pkg/package-server-manager/controller.go

```diff
@@ -0,0 +1,192 @@
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	name                = "packageserver"
+	infrastructureName  = "cluster"
+	defaultReplicaCount = int32(2)
+	defaultRolloutCount = 1
+	singleReplicaCount  = int32(1)
+)
+
+// PackageServerCSVReconciler reconciles the PackageServer CSV object
+type PackageServerCSVReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	Lock   sync.Mutex
+
+	Name                string
+	Namespace           string
+	HighlyAvailableMode bool
+}
+
+// TODO(tflannag): Determine what kubebuilder markers are required here
+func (r *PackageServerCSVReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("csv", req.NamespacedName)
+
+	log.Info("handling current request", "request", req.String())
+	log.Info("currently topology mode", "highly available", r.getHighlyAvailableMode())
+	defer log.Info("finished request reconciliation")
+
+	// TODO(tflannag): Cleanup this structure. Maybe CreateOrUpdate would help here
+	csv := &olmv1alpha1.ClusterServiceVersion{}
+	if err := r.Client.Get(ctx, req.NamespacedName, csv); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		if apierrors.IsNotFound(err) {
+			log.Info("csv does not exist attempting to create it")
+			csv, err = manifests.NewPackageServerCSV(r.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			_ = ensureCSV(log, csv, r.getHighlyAvailableMode())
+			if err := r.Client.Create(ctx, csv); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	if !csv.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("requeuing after encountering a delete event")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	orig := client.MergeFrom(csv.DeepCopy())
+	if !ensureCSV(log, csv, r.getHighlyAvailableMode()) {
+		log.Info("no updates to the csv are necessary")
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.Client.Patch(ctx, csv, orig); err != nil {
+		log.Error(err, "failed to patch the csv")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PackageServerCSVReconciler) infrastructureHandler(obj client.Object) []reconcile.Request {
+	log := r.Log.WithValues("infrastructure", obj.GetName())
+
+	if obj.GetName() != infrastructureName {
+		return nil
+	}
+
+	var infra configv1.Infrastructure
+	if err := r.Client.Get(context.Background(), types.NamespacedName{Name: obj.GetName()}, &infra); err != nil {
```

Comment:
I'm a fan of serializing this `Get` in the `Reconcile` function instead of in the infrastructure handler. I believe doing so would remove any question of how the threadiness of the manager affects consistency.